internal/provisioner/plain/controllers/bundle_controller.go

```diff
@@ -464,53 +476,186 @@ func bundleImagePod(pod *corev1.Pod, source rukpakv1alpha1.ImageSource, unpackIm
+	}
 
-	// Note: initContainer so we can ensure the repository has been cloned
-	// at the bundle.Spec.Source.Git.Ref before we unpack the Bundle contents
-	// that are stored in the repository.
-	pod.Spec.InitContainers[1].Name = "clone-repository"
-	// r.GitClientImage configures which git-based container image to use to clone the provided repository
-	// r.GitClientImage currently defaults to alpine/git:v2.32.0
-	pod.Spec.InitContainers[1].Image = gitClientImage
-	pod.Spec.InitContainers[1].ImagePullPolicy = corev1.PullIfNotPresent
-	cmd, err := git.CloneCommandFor(source)
+	userName, accessToken, err := r.getCredential(ctx, bundle)
 	if err != nil {
-		return nil, err
+		return err
+	}
+	if accessToken != "" {
+		if userName == "" {
+			userName = "user"
+		}
+		cloneOpts.Auth = &http.BasicAuth{Username: userName, Password: accessToken}
 	}
-	pod.Spec.InitContainers[1].Command = []string{"/bin/sh", "-c", cmd}
-	pod.Spec.InitContainers[1].VolumeMounts = []corev1.VolumeMount{{Name: "bundle", MountPath: "/bundle"}}
 
-	if len(pod.Spec.Containers) != 1 {
-		pod.Spec.Containers = make([]corev1.Container, 1)
+	if gitsource.Ref.Branch != "" {
+		cloneOpts.ReferenceName = plumbing.ReferenceName(fmt.Sprintf("refs/heads/%s", gitsource.Ref.Branch))
+		cloneOpts.SingleBranch = true
+		cloneOpts.Depth = 1
+	} else if gitsource.Ref.Tag != "" {
+		cloneOpts.ReferenceName = plumbing.ReferenceName(fmt.Sprintf("refs/tags/%s", gitsource.Ref.Tag))
+		cloneOpts.SingleBranch = true
+		cloneOpts.Depth = 1
 	}
 
-	pod.Spec.Containers[0].Name = bundleUnpackContainerName
-	// Note: the image for this pod is not relevant, as it exists only to run the unpacker against the bundle directory.
-	pod.Spec.Containers[0].Image = unpackImage
-	pod.Spec.Containers[0].ImagePullPolicy = corev1.PullIfNotPresent
-	pod.Spec.Containers[0].Command = []string{"/bin/unpack", "--bundle-dir", "/bundle"}
-	pod.Spec.Containers[0].VolumeMounts = []corev1.VolumeMount{{Name: "util", MountPath: "/bin"}, {Name: "bundle", MountPath: "/bundle"}}
+	// Clone
+	repo, err := git.CloneContext(ctx, memory.NewStorage(), memfs.New(), &cloneOpts)
+	if err != nil {
+		return err
+	}
+	wt, err := repo.Worktree()
+	if err != nil {
+		return fmt.Errorf("get worktree: %w", err)
+	}
 
-	pod.Spec.Volumes = []corev1.Volume{
-		{Name: "util", VolumeSource: corev1.VolumeSource{EmptyDir: &corev1.EmptyDirVolumeSource{}}},
-		{Name: "bundle", VolumeSource: corev1.VolumeSource{EmptyDir: &corev1.EmptyDirVolumeSource{}}},
+	// Checkout commit
+	if gitsource.Ref.Commit != "" {
+		commitHash := plumbing.NewHash(gitsource.Ref.Commit)
+		if err := wt.Reset(&git.ResetOptions{
+			Commit: commitHash,
+			Mode:   git.HardReset,
+		}); err != nil {
+			return fmt.Errorf("checkout commit %q: %w", commitHash.String(), err)
+		}
+	}
+
+	var bundleFS fs.FS = &billyFS{wt.Filesystem}
+
+	// Subdirectory
+	if gitsource.Directory != "" {
+		directory := gitsource.Directory
+		if directory[0] == '.' {
+			directory = directory[2:]
+		}
```

Comment:
A `.`  as the first character of a directory name doesn't always mean it will be `./`.

Can `filepath.Clean` clean things up for us? After we clean, we probably need to fail if the path starts with `../`.

Reply:
> After we clean, we probably need to fail if the path starts with ../

In fact, this seems like a thing we should check and fail on in the validation webhook.